File: ios/PaywallView.swift
//
//  PaywallView.swift
//  poool-react-native-access
//
//  Created by Ugo Stephant on 05/08/2024.
//

import UIKit
import AccessIOS

class PaywallView: UIView {
    private var access: Access!

    @objc var onLock: RCTDirectEventBlock?
    @objc var onReady: RCTDirectEventBlock?
    @objc var onRelease: RCTDirectEventBlock?
    @objc var onPaywallSeen: RCTDirectEventBlock?
    @objc var onRegister: RCTDirectEventBlock?
    @objc var onFormSubmit: RCTDirectEventBlock?
    @objc var onSubscribeClick: RCTDirectEventBlock?
    @objc var onLoginClick: RCTDirectEventBlock?
    @objc var onDiscoveryLinkClick: RCTDirectEventBlock?
    @objc var onCustomButtonClick: RCTDirectEventBlock?
    @objc var onDataPolicyClick: RCTDirectEventBlock?
    @objc var onAlternativeClick: RCTDirectEventBlock?
    @objc var onError: RCTDirectEventBlock?
    @objc var onAnswer: RCTDirectEventBlock?
    @objc var onDismissBottomSheet: RCTDirectEventBlock?

    func reinit () {
        guard let appId = appId,
              let pageType = pageType,
              let config = config,
              let styles = styles,
              let texts = texts,
              let variables = variables,
              let displayMode = displayMode else { return }

        if (access != nil) {
            access.destroy()
            access = nil
        }

        Access.setDebug(config["debug"] as? Bool ?? false)
        access = Access(key: appId)
        access.config(config)
        access.styles(styles)
        access.texts(texts)
        access.variables(variables)

        initEvents()

        switch displayMode {
            case "bottom-sheet":
            access.createPaywall(pageType: pageType, view: nil)
            default:
                let subView: UIView? = access.createPaywall(pageType: pageType)
                subView?.frame = self.frame
                
                if (subView != nil) {
                    self.addSubview(subView!)
                }
        }
    }

    private func initEvents () {
        access.onLock { self.onLock?([:]) }
        access.onReady { readyEvent in self.onReady?(readyEvent?.toMap()) }
        access.onRelease { releaseEvent in self.onRelease?(releaseEvent?.toMap()) }
        access.onPaywallSeen { seenEvent in self.onPaywallSeen?(seenEvent?.toMap()) }
        
        access.onRegister { (registerEvent, result) in
            self.onRegister?(registerEvent.toMap())
            result(nil)
        }
        access.onFormSubmit { (submitEvent, result) in
            self.onFormSubmit?(submitEvent.toMap())
            result([])
        }
        access.onSubscribeTapped { tapEvent in self.onSubscribeClick?(tapEvent?.toMap()) }
        access.onLoginTapped { tapEvent in self.onLoginClick?(tapEvent?.toMap()) }
        access.onDiscoveryLinkTapped { tapEvent in self.onDiscoveryLinkClick?(tapEvent?.toMap()) }
        access.onCustomButtonTapped { tapEvent in self.onCustomButtonClick?(tapEvent?.toMap()) }
        access.onDataPolicyTapped { tapEvent in self.onDataPolicyClick?(tapEvent?.toMap()) }
        access.onAlternativeTapEvent { tapEvent in self.onAlternativeClick?(tapEvent?.toMap()) }
        access.onError { error in self.onError?(error?.toMap()) }
        access.onAnswer { answerEvent in self.onAnswer?(answerEvent?.toMap()) }
        
        access.userDidCloseBottomSheet { self.onDismissBottomSheet?([:]) }
        
    }

    @objc var appId: String? = nil {
        didSet {
            reinit()
        }
    }

    @objc var pageType: String? = nil {
        didSet {
            reinit()
        }
    }
    
    @objc var displayMode: String? = nil {
        didSet {
            reinit()
        }
    }

    @objc var config: [String: Any]? = nil {
        didSet {
            reinit()
        }
    }

    @objc var styles: [String: Any]? = nil {
        didSet {
            reinit()
        }
    }

    @objc var texts: [String: String]? = nil {
        didSet {
            reinit()
        }
    }

    @objc var variables: [String: Any]? = nil {
        didSet {
            reinit()
        }
    }
}

File: android/src/main/java/com/poool/reactnativeaccess/PaywallView.kt
package com.poool.reactnativeaccess

import android.content.Context
import android.view.ViewGroup
import android.widget.FrameLayout
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.events.RCTEventEmitter
import tech.poool.access.Access
import tech.poool.access.onAlternativeClick
import tech.poool.access.onAnswer
import tech.poool.access.onCustomButtonClick
import tech.poool.access.onDataPolicyClick
import tech.poool.access.onDiscoveryLinkClick
import tech.poool.access.onError
import tech.poool.access.onFormSubmit
import tech.poool.access.onLock
import tech.poool.access.onLoginClick
import tech.poool.access.onPaywallSeen
import tech.poool.access.onReady
import tech.poool.access.onRegister
import tech.poool.access.onRelease
import tech.poool.access.onResize
import tech.poool.access.onSubscribeClick

class PaywallView(context: Context) : FrameLayout(context) {
  private var access: Access? = null

  private var appId: String? = null
  private var pageType: String? = null
  private var displayMode: String? = null
  private var config: Map<String, Any>? = null
  private var styles: Map<String, Any>? = null
  private var variables: Map<String, Any>? = null
  private var texts: Map<String, String>? = null

  init {
    viewTreeObserver.addOnGlobalLayoutListener { requestLayout() }
  }

  private fun reinit () {
    if (
      appId == null || pageType == null || config == null || styles == null ||
      variables == null || texts == null || displayMode == null
    ) {
      return
    }

    if (access !== null) {
      access?.destroy()
      access = null
    }

    access = Access(appId!!)
    access?.config(config!!)
    access?.styles(styles!!)
    access?.variables(variables!!)
    access?.texts(texts!!)

    initEvents()

    when (displayMode) {
      "bottom-sheet" -> {
        val ctx = context as ReactContext
        val rootView = ctx.currentActivity?.findViewById<ViewGroup>(android.R.id.content)
        val eventEmitter = ctx.getJSModule(RCTEventEmitter::class.java)

        if (rootView != null) {
          access?.createBottomSheetPaywall(pageType!!, rootView) {
            eventEmitter.receiveEvent(id, "onDismissBottomSheet", null)
          }
        }
      }

      else -> {
        val subView = access?.returnPaywallView(pageType!!, context as ThemedReactContext)
        subView?.layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
        addView(subView)
        viewTreeObserver.dispatchOnGlobalLayout()
      }
    }
  }

  fun setAppId(appId: String) {
    this.appId = appId
    reinit()
  }

  fun setPageType(pageType: String) {
    this.pageType = pageType
    reinit()
  }

  fun setDisplayMode(displayMode: String) {
    this.displayMode = displayMode
    reinit()
  }

  fun setConfig(config: ReadableMap) {
    this.config = config.toHashMap().toMap() as Map<String, Any>
    reinit()
  }

  fun setStyles(styles: ReadableMap) {
    this.styles = styles.toHashMap().toMap() as Map<String, Any>
    reinit()
  }

  fun setVariables(variables: ReadableMap) {
    this.variables = variables.toHashMap().toMap() as Map<String, Any>
    reinit()
  }

  fun setTexts(texts: ReadableMap) {
    this.texts = texts.toHashMap().toMap().mapValues { it.value.toString() }
    reinit()
  }

  private fun initEvents () {
    val reactContext = context as ReactContext
    val eventEmitter = reactContext.getJSModule(RCTEventEmitter::class.java)

    access?.onLock {
      eventEmitter.receiveEvent(id, "onLock", null)
    }

    access?.onReady {
      eventEmitter.receiveEvent(id, "onReady", PaywallEventMapping.widgetEvent(it))
    }

    access?.onRelease {
      eventEmitter.receiveEvent(id, "onRelease", PaywallEventMapping.widgetEvent(it))
    }

    access?.onPaywallSeen {
      eventEmitter.receiveEvent(id, "onPaywallSeen", PaywallEventMapping.widgetEvent(it))
    }

    access?.onRegister {
      eventEmitter.receiveEvent(id, "onRegister", PaywallEventMapping.registerEvent(it))

      return@onRegister null
    }

    access?.onFormSubmit {
      eventEmitter.receiveEvent(id, "onFormSubmit", PaywallEventMapping.formEvent(it))

      return@onFormSubmit null
    }

    access?.onSubscribeClick {
      eventEmitter.receiveEvent(id, "onSubscribeClick", PaywallEventMapping.clickEvent(it))
    }

    access?.onLoginClick {
      eventEmitter.receiveEvent(id, "onLoginClick", PaywallEventMapping.clickEvent(it))
    }

    access?.onDiscoveryLinkClick {
      eventEmitter.receiveEvent(id, "onDiscoveryLinkClick", PaywallEventMapping.clickEvent(it))
    }

    access?.onCustomButtonClick {
      eventEmitter.receiveEvent(id, "onCustomButtonClick", PaywallEventMapping.customButtonClickEvent(it))
    }

    access?.onDataPolicyClick {
      eventEmitter.receiveEvent(id, "onDataPolicyClick", PaywallEventMapping.clickEvent(it))
    }

    access?.onAlternativeClick {
      eventEmitter.receiveEvent(id, "onAlternativeClick", PaywallEventMapping.alternativeClickEvent(it))
    }

    access?.onError { e, _ ->
      eventEmitter.receiveEvent(id, "onError", PaywallEventMapping.errorEvent(e))
    }

    access?.onAnswer {
      eventEmitter.receiveEvent(id, "onAnswer", PaywallEventMapping.answerEvent(it))
    }

    access?.onResize {
      eventEmitter.receiveEvent(id, "onResize", PaywallEventMapping.resizeEvent(it))
    }
  }
}
